Draft-Accepted/RFCXXXX-Command-Shell-Integration.md

```diff
@@ -0,0 +1,287 @@
+    "executable": "az",
+    "version": "1.0.0"
+}
+```
+
+The `$schema` includes the version of the manifest with the URL pointing to a published JSON schema.
+The `executable` is the name of the executable that the manifest is for and expected to be found in `$env:PATH`
+The `version` is the version of the manifest for the tool and only used for informational purposes.
+
+### Structured Output and Custom Formatting Registration
+
+Tools can indicate they can output [JSONLines](https://jsonlines.org/) by default and also provide custom formatting.
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "structuredOutput": {
+        "default": "jsonlines"
+    }
+}
+```
+
+If a tool indicates it can output structured data in a format supported by the shell,
+the shell will set the env var `COMMAND_SHELL_STRUCTURED_OUTPUT` with the value `jsonlines` in this case
+in the environment block of the process that is launched.
+
+Tools should check for this environment variable and output structured data in the format indicated if supported.
+
+If the tool calls other executables, it is responsible for removing or propagating the environment variable which
+can affect the behavior of child processes.
+
+JSON output should include a `$typeNames` member which is an array of strings that indicate the type hierarchy of the object.
+PowerShell would use this as the same as the `TypeNames` member of a PSObject for formatting.
+Future enhancement would support a JSON defined formatting schema that could be used by other shells.
+
+### Tab Completion Registration
+
+Within the same manifest, a tool can indicate it supports tab completion.
+
+This can either be an executable with arguments.
+
+Arguments can be built-in variables to provide context for the tab completion.
+
+Currently supported variables:
+
+- `{commandLine}` - The full command line that the user has typed so far.
+- `{cursorPosition}` - The current cursor position within the command line, starting at 0 index.
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "tabCompletion": {
+        "command": {
+            "executable": "az",
+            "arguments": [
+                "completion",
+                "powershell",
+                "--commandLine",
+                "{commandLine}",
+                "--cursorPosition",
+                "{cursorPosition}"
+            ]
+        }
+    }
+}
+```
+
+or a shell specific script:
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "tabCompletion": {
+        "script": {
+            "powershell": "az-completion.ps1",
```

Comment:
Does this mean calling `az-completion.ps1` for every tab completion triggered for `az`, or is it a one-time thing for registration?

Reply:
Good question.  I think it's shell specific.  In the PowerShell case, I'd expect it to be one time registration.  Bash probably will use it as the completer each time (I'm sure bash isn't going to change their current model, this is just an example).